src/Development/IDE/Core/Shake.hs

```diff
@@ -377,57 +395,115 @@ shakeProfile :: IdeState -> FilePath -> IO ()
+
+-- Set up a new 'ShakeSession' with a set of initial actions.
+-- Will crash if there is an existing 'ShakeSession' running.
+-- Reports progress on the set of initial actions only.
+newSession :: IdeState -> [Action a] -> IO (ShakeSession, IO [a])
+newSession IdeState{shakeExtras=ShakeExtras{..}, ..} acts = do
+    -- A work queue for actions added via 'runInShakeSession'
+    actionQueue :: TQueue (Action ()) <- newTQueueIO
+
+    start <- offsetTime
+    let
+        -- A daemon-like action used to inject additional work
+        -- Runs actions from the work queue sequentially
+        pumpAction =
+            forever $ join $ liftIO $ atomically $ readTQueue actionQueue
+
+        progressRun
+          | reportProgress = lspShakeProgress ideTesting getLspId eventer inProgress
+          | otherwise = return ()
+
+        workRun restore = withAsync progressRun $ \progressThread -> do
+          let acts' =
+                [ [] <$ pumpAction
+                , parallel acts
+                    -- For the purposes of progress reporting, only 'acts' matters.
+                    -- When done, cancel the progressThread to indicate completion
+                    <* liftIO (cancel progressThread)
+                ]
+          res <- try (restore $ shakeRunDatabaseProfile shakeProfileDir shakeDb acts')
+          runTime <- start
+          let res' = case res of
+                      Left e -> "exception: " <> displayException e
+                      Right _ -> "completed"
+              profile = case res of
+                      Right (_, Just fp) ->
+                          let link = case filePathToUri' $ toNormalizedFilePath' fp of
+                                          NormalizedUri _ x -> x
+                          in ", profile saved at " <> T.unpack link
+                      _ -> ""
+              -- Wrap up in a thread to avoid calling interruptible
+              -- operations inside the masked section
+          let wrapUp = do
+                  logDebug logger $ T.pack $
+                      "Finishing shakeRun (took " ++ showDuration runTime ++ ", " ++ res' ++ profile ++ ")"
+          return (fst <$> res, wrapUp)
+
+    -- Do the work in a background thread
+    workThread <- asyncWithUnmask workRun
+    -- run the wrap up unmasked
+    _ <- async $ do
+        (_, wrapUp) <- wait workThread
+        wrapUp
+
+    -- 'runInShakeSession' is used to append work in this Shake session
+    --  The session stays open until 'cancelShakeSession' is called
+    --  This should only be necessary iff the (virtual) filesystem has changed
+    let runInShakeSession :: forall a . Action a -> IO (IO a)
+        runInShakeSession act = do
+              res <- newEmptyMVar
+              let act' = actionCatch @SomeException (Right <$> act) (pure . Left)
+              atomically $ writeTQueue actionQueue (act' >>= liftIO . putMVar res)
+              return (takeMVar res >>= either throwIO return )
```

Comment:
I would have used a barrier here for res - makes it clear the MVar is usually empty and then written to once. That might just be my style though (I hate that MVar's mean one of 3 distinct things depending on their usage, so always pick one of my three wrappers - https://neilmitchell.blogspot.com/2012/06/flavours-of-mvar_04.html

Reply:
I never saw the point of those concurrency abstractions until now, thanks for the link! Now I understand where they are coming from and see the value in them. 

That said, when I look at the implementation details, I'm back firmly in the `MVar` camp:
```
newtype Barrier a = Barrier (Var (Either (MVar ()) a))
```
I'm just too terrified of concurrent code to trust something like that :sweat_smile:

Reply:
I should fix that. I wanted the semantics that you could only signal once, but you could wait multiple times. With older GHC that was impossible to achieve directly on MVar. Now it is possible I'll put it back to being a nice simple MVar wrapper.

Reply:
See https://github.com/ndmitchell/extra/commit/98c2a83585d2ca0a9d961dd241c4a967ef87866a - now all the concurrency wrappers are very simple zero-cost zero-complexity over MVar - but guarantee that you can't leave it empty (Var), or can't fill it more than once (Barrier).

Reply:
That looks much better, and I can happily switch to a Barrier now